paymentsheet/src/main/java/com/stripe/android/customersheet/data/CustomerAdapterDataSource.kt

```diff
@@ -1,24 +1,71 @@
 package com.stripe.android.customersheet.data
 
+import com.stripe.android.common.coroutines.runCatching
+import com.stripe.android.core.exception.StripeException
+import com.stripe.android.core.injection.IOContext
 import com.stripe.android.customersheet.CustomerAdapter
 import com.stripe.android.customersheet.CustomerAdapter.PaymentOption.Companion.toPaymentOption
+import com.stripe.android.customersheet.CustomerPermissions
 import com.stripe.android.customersheet.ExperimentalCustomerSheetApi
 import com.stripe.android.customersheet.map
+import com.stripe.android.lpmfoundations.paymentmethod.PaymentMethodSaveConsentBehavior
+import com.stripe.android.model.ElementsSession
 import com.stripe.android.model.PaymentMethod
 import com.stripe.android.model.PaymentMethodUpdateParams
+import com.stripe.android.payments.core.analytics.ErrorReporter
+import com.stripe.android.paymentsheet.PaymentSheet
 import com.stripe.android.paymentsheet.model.SavedSelection
+import com.stripe.android.paymentsheet.repositories.ElementsSessionRepository
+import kotlinx.coroutines.async
 import javax.inject.Inject
 import javax.inject.Singleton
+import kotlin.coroutines.CoroutineContext
 
 @OptIn(ExperimentalCustomerSheetApi::class)
 @Singleton
 internal class CustomerAdapterDataSource @Inject constructor(
+    private val elementsSessionRepository: ElementsSessionRepository,
     private val customerAdapter: CustomerAdapter,
-) : CustomerSheetSavedSelectionDataSource,
+    private val errorReporter: ErrorReporter,
+    @IOContext private val workContext: CoroutineContext,
+) : CustomerSheetInitializationDataSource,
+    CustomerSheetSavedSelectionDataSource,
     CustomerSheetPaymentMethodDataSource,
     CustomerSheetIntentDataSource {
     override val canCreateSetupIntents: Boolean = customerAdapter.canCreateSetupIntents
 
+    override suspend fun loadCustomerSheetSession(): CustomerSheetDataResult<CustomerSheetSession> {
+        return workContext.runCatching {
+            val elementsSessionResult = async {
+                fetchElementsSession()
+            }
+
+            val paymentMethodsResult = async {
+                fetchInitialPaymentMethods()
+            }
+
+            val savedSelectionResult = async {
+                retrieveSavedSelection().toResult()
+            }
+
+            val elementsSession = elementsSessionResult.await().getOrThrow()
```

Comment:
is getOrThrow the right thing here or can we use a fallback? Presumably this would cause a crash if someone had bad connectivity and the network request failed?

Reply:
All the `getOrThrow` calls here are wrapped with `runCatching` so that if something is thrown, it's caught and returned as a `Result.failure`

Reply:
There are also tests to ensure these are caught by `runCatching`.